src/vmspawn/vmspawn-util.c

```diff
@@ -0,0 +1,195 @@
+                _cleanup_close_ int fd = -EBADF;
+                fd = open(firmware, O_CLOEXEC | O_RDONLY);
+                if (fd >= 0) {
+                        if (ret_firmware_path)
+                                *ret_firmware_path = firmware;
+                        return 0;
+                }
+        }
+
+#undef FIRMWARE_LOCATIONS
+        #define FIRMWARE_LOCATIONS \
+                "/usr/share/edk2/ovmf/OVMF_CODE.fd", \
+                "/usr/share/edk2-ovmf/OVMF_CODE.fd", \
+                "/usr/share/qemu/OVMF_CODE.fd", \
+                "/usr/share/ovmf/OVMF.fd", \
+                "/usr/share/OVMF/OVMF_CODE.fd"
+
+        FOREACH_STRING(firmware, FIRMWARE_LOCATIONS) {
+                _cleanup_close_ int fd = -EBADF;
+                fd = open(firmware, O_CLOEXEC | O_RDONLY);
+                if (fd >= 0) {
+                        log_warning("Couldn't find OVMF firmware blob with secure boot support, "
+                                    "falling back to OVMF firmware blobs without secure boot support.");
+
+                        if (ret_firmware_path)
+                                *ret_firmware_path = firmware;
+                        return 1;
+                }
+        }
+
+#undef FIRMWARE_LOCATIONS
+        if (ret_firmware_path)
+                *ret_firmware_path = NULL;
+        log_error("Couldn't find OVMF UEFI firmware blob.");
+        return -ENOENT;
+}
+
+int find_qemu_binary(char **ret_qemu_binary) {
+        int r;
+
+        static const char *architecture_to_qemu_table[_ARCHITECTURE_MAX] = {
+                [ARCHITECTURE_ARM64]       = "aarch64",
+                [ARCHITECTURE_ARM]         = "arm",
+                [ARCHITECTURE_ALPHA]       = "alpha",
+                [ARCHITECTURE_X86_64]      = "x86_64",
+                [ARCHITECTURE_X86]         = "i386",
+                [ARCHITECTURE_LOONGARCH64] = "loongarch64",
+                [ARCHITECTURE_MIPS64_LE]   = "mips",
+                [ARCHITECTURE_MIPS_LE]     = "mips",
+                [ARCHITECTURE_PARISC]      = "hppa",
+                [ARCHITECTURE_PPC64_LE]    = "ppc",
+                [ARCHITECTURE_PPC64]       = "ppc",
+                [ARCHITECTURE_PPC]         = "ppc",
+                [ARCHITECTURE_RISCV32]     = "riscv32",
+                [ARCHITECTURE_RISCV64]     = "riscv64",
+                [ARCHITECTURE_S390X]       = "s390x",
+        };
+
+        FOREACH_STRING(s, "qemu", "qemu-kvm") {
+                r = find_executable(s, ret_qemu_binary);
+                if (r == 0)
+                        return 0;
+        }
+
+        const char *arch_qemu = architecture_to_qemu_table[native_architecture()];
+        if (!arch_qemu) {
+                log_error("Architecture %s not supported by qemu", architecture_to_string(native_architecture()));
+                return -ENOENT;
+        }
+
+        _cleanup_free_ char *qemu_arch_specific = NULL;
+        asprintf(&qemu_arch_specific, "qemu-system-%s", arch_qemu);
```

Comment:
oom check missing.

also, please use strjoin() if all you need to do is concat a bunch of strings. sprintf() is terribly slow in comparisons

Reply:
Fixed in 111681abaf.